sw/device/silicon_creator/lib/shutdown.c

```diff
@@ -255,6 +257,60 @@ shutdown_error_redact_t shutdown_redact_policy(void) {
   return shutdown_redact_policy_inline(raw_state);
 }
 
+enum {
+  /**
+   * Length of the error message prefix.
+   */
+  kMsgPrefixLen = 14,
+  /**
+   * Length of the hexadecimal representation.
+   */
+  kHexCharsLen = 8,
+  /**
+   * Base address of UART.
+   */
+  kUartBase = TOP_EARLGREY_UART0_BASE_ADDR,
+};
+
+static_assert((kMsgPrefixLen + kHexCharsLen) <= 32,
+              "Total message length must be less than TX FIFO size.");
+
+/**
+ * Prints a fixed-length message prefix.
+ *
+ * @param prefix Prefix to print.
+ */
+static ALWAYS_INLINE void shutdown_print_prefix(const char *prefix) {
+  for (size_t i = 0; i < kMsgPrefixLen; ++i) {
+    // `kMsgPrefixLen` is laundered so that it is loaded to a register at every
+    // iteration.
+    if (i >= launder32(kMsgPrefixLen)) {
```

Comment:
Why is this necessary? Can't you just do this comparison in the loop? Is this intended to be a funny hardening?

Reply:
Kind of. The loop already has this check but this is meant to be a redundant check against printing more than we want to, loading `kMsgPrefixLen` at each iteration. PTAL at `bc56` below:

```
launder32():
/workspace/worktrees/opentitan/boot-policy/sw/device/lib/base/hardened.h:242
    bc56:	47b9                	li	a5,14
shutdown_print_prefix():
/workspace/worktrees/opentitan/boot-policy/build-out/../sw/device/silicon_creator/lib/shutdown.c:287
    if (i >= launder32(kMsgPrefixLen)) {
    bc58:	00f57a63          	bgeu	a0,a5,bc6c <shutdown_finalize+0x1bc>
/workspace/worktrees/opentitan/boot-policy/build-out/../sw/device/silicon_creator/lib/shutdown.c:290
    abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, prefix[i]);
    bc5c:	00d507b3          	add	a5,a0,a3
    bc60:	0007c783          	lbu	a5,0(a5)
/workspace/worktrees/opentitan/boot-policy/build-out/../sw/device/silicon_creator/lib/shutdown.c:284
  for (size_t i = 0; i < kMsgPrefixLen; ++i) {
    bc64:	0505                	addi	a0,a0,1
abs_mmio_write32():
/workspace/worktrees/opentitan/boot-policy/sw/device/lib/base/abs_mmio.h:91
  *((volatile uint32_t *)addr) = value;
    bc66:	cf5c                	sw	a5,28(a4)
shutdown_print_prefix():
/workspace/worktrees/opentitan/boot-policy/build-out/../sw/device/silicon_creator/lib/shutdown.c:284
    bc68:	feb517e3          	bne	a0,a1,bc56 <shutdown_finalize+0x1a6>
```